Add Check Support for Java 21 Unnamed Variables & Patterns Syntax: AnnotationLocation

child of #14942 :
Check documentation:https://checkstyle.org/checks/annotation/annotationlocation.html#AnnotationLocation

```
PS D:\CS\test> cat src/Test.java                                                
record ColoredPoint(int p, int x, int c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }

public class Test {
    void test(Object obj) {
        if (obj instanceof ColoredPoint( @Deprecated int _,_,_)){
        }
        @Deprecated
        int _ = 0;
    }
}

PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="AnnotationLocation">
        </module>
    </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
Audit done.
PS D:\CS\test> 
```

---
No update is needed for this check.

I noticed that the variables inside the methods are not being checked. this check only checks for class fields if this is the correct 
behavior (nothing in the doc says that so I am not sure but all examples are on class fields) then we are good we can't have unnamed fields 
```
PS D:\CS\test> cat src/Test.java                                                
record ColoredPoint(int p, int x, String c) { } 
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }
public class Test {
    @SuppressWarnings("deprecation") int x = 0;         // violation
    void test(Object obj) {
        if (obj instanceof ColoredPoint(@SuppressWarnings("deprecation") int _, _ , _)) {
        }
        @SuppressWarnings("deprecation") int x = 0;  // ok ?
    }

}
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:4:5: Annotation 'SuppressWarnings' should be alone on line. [AnnotationLocation]
Audit done.
Checkstyle ends with 1 errors.
PS D:\CS\test> 

```

---

> Can the annotation itself have an unnamed?

nope.

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationLocationCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationLocationCheckTest.java
@@ -331,4 +331,12 @@ public class AnnotationLocationCheckTest extends AbstractModuleTestSupport {
             expected);
     }
 
+    @Test
+    public void testAnnotationLocationOnLocalAndPatternVariables() throws Exception {
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputAnnotationLocationLocalAndPatternVariables.java"),
+            expected);
+    }
+
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/annotation/annotationlocation/InputAnnotationLocationLocalAndPatternVariables.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/annotation/annotationlocation/InputAnnotationLocationLocalAndPatternVariables.java
@@ -1,0 +1,29 @@
+/*
+AnnotationLocation
+allowSamelineMultipleAnnotations = (default)false
+allowSamelineSingleParameterlessAnnotation = (default)true
+allowSamelineParameterizedAnnotation = (default)false
+tokens = (default)CLASS_DEF, INTERFACE_DEF, PACKAGE_DEF, ENUM_CONSTANT_DEF, \
+         ENUM_DEF, METHOD_DEF, CTOR_DEF, VARIABLE_DEF, RECORD_DEF, COMPACT_CTOR_DEF
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.annotation.annotationlocation;
+
+public class InputAnnotationLocationLocalAndPatternVariables {
+
+    void test(Object obj) {
+       @SuppressWarnings("deprecation") int x = 5;
+       @SuppressWarnings("deprecation") int _ = 5;
+       if (obj instanceof ColoredPoint( @Deprecated int y,_,_)) {}
+       if (obj instanceof ColoredPoint( @Deprecated int _,_,_)) {}
+
+       switch (obj) {
+           case ColoredPoint( @Deprecated int y,_,_) -> {}
+           default -> {}
+       }
+    }
+    record ColoredPoint(int p, int x, int c) { }
+}

